pkgs/top-level/default.nix

```diff
@@ -22,16 +22,18 @@
   # `*Platform`s.
   localSystem
 
-, # The system packages will ultimately be run on. Null if the two should be the
-  # same.
-  crossSystem ? null
+, # The system packages will ultimately be run on.
+  crossSystem ? localSystem
 
 , # Allow a configuration attribute set to be passed in as an argument.
   config ? {}
 
 , # List of overlays layers used to extend Nixpkgs.
   overlays ? []
 
+, # List of overlays to apply to target packages only.
+  crossOverlays ? []
```

Comment:
This looks like a great improvement. As @Ericson2314 mentioned, this brings us much closer to `build` and `host` terminology. So what about changing this to `hostOverlays`, because then it can be applied to non-cross as well.

Reply:
With non cross, the build and host packages are the same, so it's not possible to just override in one case. `crossOverrides` actually will turn on cross if no `crossSystem` (to be renamed :)) was given to make things work.

Reply:
One reason that crossOverlays makes sense is that you will be using the "cross" stdenv here. This could be a little surprising, for instance, if you use crossOverlays on macOS, you will get CC=gcc instead of CC=clang.

Reply:
@matthewbauer so that's the `host` or `run` `CC`, right? So why not `hostOverlays` or `runOverlays`? I'd say "cross" is a property of the entire package set: was it cross-compiled? yes or no, boolean. When describing which architecture some variable applies to, it should be `build` or `host`/`run`, because that is precise.

Now I might be entirely wrong about this, but this, `crossOverlays`, seems very similar to the way in which I accidentally wrote `pkgs.extend` to be [broken](https://github.com/NixOS/nixpkgs/commit/889a5c35127e4e6b2c722611dcaa9c802337446b#commitcomment-31560635).

I'm trying to do something at the edge of my understanding of Nixpkgs internals here on a limited time budget. My initial goal was to achieve greater consistency, mainly between the Nixpkgs function and NixOS, so that it will be as simple as possible to understand the interface and work with it. I hope that a new-ish perspective is useful; if not, I should focus on other things.

Reply:
@robeth That looks like okay logic. We definitely want to unify the build, host, target & local, cross, eventually. But I still am thinking crossOverlays is good here because it mirrors crossSystem. Right now, this is how I see it-

- {build,host,target}Platform - used internally in Nixpkgs as the "immediate" values of build, host, target. These are given to you, not set directly. Sliding windows algorithm handles the details...
- {local,cross}System - external arguments of Nixpkgs describing **where** you are and where you want to go. This is what you want to set to get from point a to point b.

If cross-compiling is a graph, then each node is a set of packages. We activate a kind of greedy pathfinding algorithm in moving from localSystem to crossSystem. Outside of Nixpkgs, we shouldn't care about the exact path that is taken, just that we end up with a package built by "localSystem", for "crossSystem". In the same way, when you set "crossOverlays", you are saying to give me a package set that is statically built, but using my localSystem to avoid rebooting everything statically.